crates/nu-command/src/network/fetch.rs

```diff
@@ -0,0 +1,386 @@
+};
+
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::time::Duration;
+
+#[derive(Clone)]
+pub struct SubCommand;
+
+impl Command for SubCommand {
+    fn name(&self) -> &str {
+        "fetch"
+    }
+
+    fn signature(&self) -> Signature {
+        Signature::build("fetch")
+            .desc("Load from a URL into a cell, convert to table if possible (avoid by appending '--raw').")
+            .required(
+                "URL",
+                SyntaxShape::String,
+                "the URL to fetch the contents from",
+            )
+            .named(
+                "user",
+                SyntaxShape::Any,
+                "the username when authenticating",
+                Some('u'),
+            )
+            .named(
+                "password",
+                SyntaxShape::Any,
+                "the password when authenticating",
+                Some('p'),
+            )
+            .named("timeout", SyntaxShape::Int, "timeout period in seconds", Some('t'))
+            .switch("raw", "fetch contents as text rather than a table", Some('r'))
+            
+            .filter()
+            .category(Category::Network)
+    }
+
+    fn usage(&self) -> &str {
+        "Fetch the contents from a URL (HTTP GET operation)."
+    }
+
+    fn run(
+        &self,
+        engine_state: &EngineState,
+        stack: &mut Stack,
+        call: &Call,
+        input: PipelineData,
+    ) -> Result<nu_protocol::PipelineData, nu_protocol::ShellError> {
+        run_fetch(engine_state, stack, call, input)
+    }
+
+    fn examples(&self) -> Vec<Example> {
+        vec![
+            Example {
+                description: "Fetch content from url.com",
+                example: "fetch url.com",
+                result: None,
+            },
+            Example {
+                description: "Fetch content from url.com, with username and password",
+                example: "fetch -u myuser -p mypass url.com",
+                result: None,
+            },
+        ]
+    }
+}
+
+fn run_fetch(
+    engine_state: &EngineState,
+    stack: &mut Stack,
+    call: &Call,
+    _input: PipelineData,
+) -> Result<nu_protocol::PipelineData, nu_protocol::ShellError> {
+    let mut fetch_helper = Fetch::new();
+
+    fetch_helper.setup(engine_state, call, stack)?;
+
+    let runtime = tokio::runtime::Runtime::new()?;
+
+    // let path = match fetch_helper.path {
+    //     Some(p) => p,
+    //     None => {
+    //         return Err(ShellError::UnsupportedInput(
+    //             "The url must be a string".to_string(),
+    //             call.head,
+    //         ))
+    //     }
+    // };
+
+    runtime.block_on(fetch(
+        engine_state,
+        call,
+        stack,
+        fetch_helper.url,
+        fetch_helper.has_raw,
+        fetch_helper.user.clone(),
+        fetch_helper.password,
+        fetch_helper.timeout,
+    ))
+}
+
+#[derive(Default)]
+pub struct Fetch {
```

Comment:
Hi! @onthebridgetonowhere 
I'm interested in fetch and post porting, so actually I've seen the original nushell code.

Personally, I wonder if the `Fetch` struct is really necessary.
This simply takes arguments, but I think it's too much for the purpose.

So, How about an approach like this? (I wrote it rough)
```rust
use use base64::encode;
...

struct Args {
  // url, has_raw, user ..
}

...

fn run_fetch( ... ) {
  let args: Args = {
    url: Some(call.req(engine_state, stack, 0)?);
    ...
  }

  ...

  fetch(engine_state, call, stack, args}
}
```

Reply:
Thanks for the suggestions @jaeheonji. I guess we can do that; it's sometimes confusing how to handle things as we have several styles going on at the moment, so it would be great to unify them (e.g., keep only the struct Args for different arguments). I mostly followed what was done in the original nushell fetch command and focused on getting it to work, without refactoring or simplifying the code. I am still working on the byte stream part, but feel free to suggest changes :-)

Reply:
I've been cleaning this up. What do you think @jaeheonji ?

Reply:
LGMT!